### PR TITLE
Add permutation functions

### LINKED ISSUE #1319: Add next_permutation, prev_permutation, is_permutation
Add the permutation algorithms

### PR DESCRIPTION
**Pull request by wsciaroni**
This pull request adds support for permutation generation algorithms.

**Algorithm improvements and new features:**

* Improved the `etl::is_permutation` implementation to use `count_if` with binary predicates, ensuring correct behavior when custom predicates are provided, and added explicit checks for range length mismatches.
* Added implementations for `etl::next_permutation` and `etl::prev_permutation`, both with and without custom comparison functions, matching standard library behavior.

**Testing enhancements:**

* Added tests for `next_permutation` and `prev_permutation` (including custom comparators and edge cases), as well as new tests for `is_permutation` with length mismatches and custom predicates.
* Introduced a new `DataEquivalenceByA` predicate struct to support custom predicate testing for `is_permutation`.

Closes #1319

### CODE REVIEW

File: include/etl/algorithm.h [hunk 1]
```diff
@@ -1973,6 +1983,206 @@ namespace etl
...
   }
 
+  //***************************************************************************
+  /// next_permutation
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/next_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator>
+  ETL_CONSTEXPR14
+  bool next_permutation(TIterator first, TIterator last)
```
Comment (jwellbelove) [inline include/etl/algorithm.h]: This could be implemented by calling the predicate version of next_permutation with etl::less<typename etl::iterator_traits<TIterator>::value_type>

File: include/etl/algorithm.h [hunk 2]
```diff
@@ -1973,6 +1983,206 @@ namespace etl
...
+  }
+
+  //***************************************************************************
+  /// prev_permutation
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/prev_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator>
+  ETL_CONSTEXPR14
+  bool prev_permutation(TIterator first, TIterator last)
```
Comment (jwellbelove) [inline include/etl/algorithm.h]: This could be implemented by calling the predicate version of prev_permutation with etl::less<typename etl::iterator_traits<TIterator>::value_type>

File: include/etl/algorithm.h (hunk 1 above)
Comment (wsciaroni) [reply include/etl/algorithm.h]: I have updated this implementation to invoke the other function with `etl::less<...`.

File: include/etl/algorithm.h (hunk 2 above)
Comment (wsciaroni) [reply include/etl/algorithm.h]: I have updated this implementation to invoke the other function with `etl::less<...`.

File: test/test_algorithm.cpp [hunk 3]
```diff
@@ -1318,6 +1326,121 @@ namespace
...
+    TEST(next_permutation)
+    {
+      std::array<int, 4U> expected = { 1, 1, 2, 2 };
+      std::array<int, 4U> result = expected;
+
+      for (size_t i = 0U; i < 8U; ++i)
+      {
+        bool expected_has_next = std::next_permutation(expected.begin(), expected.end());
+        bool result_has_next = etl::next_permutation(result.begin(), result.end());
+
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply test/test_algorithm.cpp]: Adding a test for when the beginning is the same as the end.

File: include/etl/algorithm.h [hunk 4]
```diff
@@ -1973,6 +1983,132 @@ namespace etl
     return true;
   }
 
+  //***************************************************************************
+  /// next_permutation
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/next_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator, typename TCompare>
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h:2105]: This is not the interface of the std:: implementation, so we will not be adding nodiscard to this method.

File: include/etl/algorithm.h [hunk 5]
```diff
@@ -1973,6 +1983,132 @@ namespace etl
...
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/next_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator>
+  ETL_CONSTEXPR14
+  bool next_permutation(TIterator first, TIterator last)
+  {
+    typedef etl::less<typename etl::iterator_traits<TIterator>::value_type> compare;
+    return etl::next_permutation(first, last, compare());
+  }
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h:2161]: This is not the interface of the std:: implementation, so we will not be adding nodiscard to this method.

File: include/etl/algorithm.h [hunk 6]
```diff
@@ -1973,6 +1983,132 @@ namespace etl
...
+
+  //***************************************************************************
+  /// prev_permutation
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/prev_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator, typename TCompare>
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h:2168]: This is not the interface of the std:: implementation, so we will not be adding nodiscard to this method.

File: include/etl/algorithm.h [hunk 7]
```diff
@@ -1973,6 +1983,132 @@ namespace etl
...
+  ///\ingroup algorithm
+  ///<a href="http://en.cppreference.com/w/cpp/algorithm/prev_permutation"></a>
+  //***************************************************************************
+  template <typename TIterator>
+  ETL_CONSTEXPR14
+  bool prev_permutation(TIterator first, TIterator last)
+  {
+    typedef etl::less<typename etl::iterator_traits<TIterator>::value_type> compare;
+    return etl::prev_permutation(first, last, compare());
+  }
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h:2224]: This is not the interface of the std:: implementation, so we will not be adding nodiscard to this method.

File: include/etl/algorithm.h [hunk 8]
```diff
@@ -1896,9 +1896,9 @@ namespace etl
       {
         if (i == etl::find_if(begin1, i, etl::bind1st(predicate, *i)))
         {
-          size_t n = etl::count(begin2, end2, *i);
+          size_t n = etl::count_if(begin2, end2, etl::bind1st(predicate, *i));
 
-          if (n == 0 || size_t(etl::count(i, end1, *i)) != n)
+          if (n == 0 || size_t(etl::count_if(i, end1, etl::bind1st(predicate, *i))) != n)
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h]: Valid change. I implemented without the use of auto to preserve 98 compatibility.

File: include/etl/algorithm.h [hunk 9]
```diff
@@ -1954,15 +1959,20 @@ namespace etl
...
     {
       for (TIterator1 i = begin1; i != end1; ++i)
       {
         if (i == etl::find_if(begin1, i, etl::bind1st(predicate, *i)))
         {
-          size_t n = etl::count(begin2, end2, *i);
+          size_t n = etl::count_if(begin2, end2, etl::bind1st(predicate, *i));
 
-          if (n == 0 || size_t(etl::count(i, end1, *i)) != n)
+          if (n == 0 || size_t(etl::count_if(i, end1, etl::bind1st(predicate, *i))) != n)
```
[the remark this answers is not in the document]
Comment (wsciaroni) [reply include/etl/algorithm.h]: Valid change. I implemented without the use of auto to preserve 98 compatibility.

File: include/etl/algorithm.h [hunk 10]
```diff
@@ -1894,11 +1894,12 @@ namespace etl
 
       for (TIterator1 i = begin1; i != end1; ++i)
       {
-        if (i == etl::find_if(begin1, i, etl::bind1st(predicate, *i)))
+        const typename etl::binder1st<TBinaryPredicate> predicate_is_i = etl::bind1st(predicate, *i);
+        if (i == etl::find_if(begin1, i, predicate_is_i))
         {
-          size_t n = etl::count(begin2, end2, *i);
+          size_t n = etl::count_if(begin2, end2, predicate_is_i);
```
Comment (wsciaroni) [inline include/etl/algorithm.h:2011]: This change in `is_permutation` requires callers to make the predicate an `etl::binary_function`. Is that acceptable @jwellbelove

Comment (wsciaroni) [reply include/etl/algorithm.h:2011]: @rolandreichweinbmw thoughts?

Comment (rolandreichweinbmw) [reply include/etl/algorithm.h:2011]: I agree that this is the obvious fix.
